[Instrumentation] Do not run sanitizers for naked functions

antoniofrighetto opened the pull request:
Sanitizers instrumentation may be incompatible with naked functions, which lack of standard prologue/epilogue.

fmayer commented:
What about HWAddressSanitizer?

antoniofrighetto commented:
> What about HWAddressSanitizer?

Added!

fmayer commented:
I am not sure the commit message is completely accurate though. IR docs say

> This attribute disables prologue / epilogue emission for the function. This can have very system-specific consequences. The arguments of a naked function can not be referenced through IR values.

This means _non arguments_ could still be referenced through IR.

It would probably help if you could point towards a specific problem that this is solving.

antoniofrighetto commented:
See downstream issue: https://github.com/rust-lang/rust/issues/129224. Yeah, the commit message can be made more accurate. I was having in mind Rust Naked functions when I wrote it; indeed inline asm only is not a strict requirement in LLVM. That said, I'm not sure you'd be able to leverage Asan instrumentation properly (e.g., for stack accesses), if the function lacks of prologue/epilogue in the first place.

fmayer commented:
LGTM % a more accurate commit message

vitalybuka commented:
I suspect msan/dfsan still needs to set shadow for output value.
asan/hwasan probably has nothing to touch a naked function anyway.

vitalybuka commented:
also msan instruments inline assembly, I think we still need it for naked function in general

vitalybuka commented on llvm/test/Instrumentation/sanitizers-no-naked.ll:
```diff
@@ -0,0 +1,60 @@
+; NOTE: Assertions have been autogenerated by utils/update_test_checks.py UTC_ARGS: --version 5
+; RUN: opt < %s -passes=asan -S | FileCheck %s
+; RUN: opt < %s -passes=tsan -S | FileCheck %s
+; RUN: opt < %s -passes=hwasan -S | FileCheck %s --check-prefixes=CHECK-HWASAN
```
there is no `-passes=msan`

antoniofrighetto commented:
Tests in a separate commit, rephrased commit, dropped msan/dfsan. Does this look better?

nikic commented on llvm/test/Instrumentation/sanitizers-naked.ll:
```diff
@@ -0,0 +1,90 @@
+target datalayout = "e-m:e-i64:64-f80:128-n8:16:32:64-S128"
+target triple = "x86_64-unknown-linux-gnu"
+
+define void @naked_function() naked {
```
Can you please add a byval argument to these functions? I think that will make it more likely that something actually gets instrumented and results in a verifier error.

antoniofrighetto replied:
May be missing something, how is the argument supposed to lead that something gets instrumented if the argument cannot be used?

nikic replied:
See https://llvm.godbolt.org/z/K4WdqTdve. Without the byval argument there is nothing to instrument anyway, so you're not really testing anything.

antoniofrighetto replied:
Doesn't your function miss naked attribute? The verifier would lead to an error in that case.

nikic replied:
Yes, that's the point. You want a test case that leads to a verifier error before your change and no longer causes one after your change.

antoniofrighetto replied:
OK I see it now, thanks. Should be better. I was slightly confused by the naked_function without being naked.

nikic commented on llvm/test/Instrumentation/sanitizers-naked.ll:
```diff
@@ -0,0 +1,89 @@
+; NOTE: Assertions have been autogenerated by utils/update_test_checks.py UTC_ARGS: --version 5
+; RUN: opt < %s -passes=asan -S | FileCheck %s --check-prefixes=CHECK-ASAN
```
```suggestion
; RUN: opt < %s -passes=asan -S | FileCheck %s --check-prefixes=CHECK,CHECK-ASAN
```
etc. to remove duplicate check lines.

antoniofrighetto replied:
Added, but they seem unused.

nikic replied:
Hm okay, sounds like a bug in UTC.

antoniofrighetto committed:
[Instrumentation] Do not request sanitizers for naked functions

Sanitizers instrumentation may be incompatible with naked functions,
which lack of standard prologue/epilogue.

antoniofrighetto commented:
@nikic Think this is OK to land for you too?

nikic reviewed (approved):
LGTM

antoniofrighetto commented:
Thanks!
